[ChinZJ] iP

ChinZJ opened the pull request:
# Botling  
> "Always code as if the guy who ends up maintaining your code will be a violent psychopath who knows where you live." - John Woods [(source)](https://www.goodreads.com/quotes/248194-always-code-as-if-the-guy-who-ends-up-maintaining)  
  
Botling is a task list application that is     
* text-based
* easy to learn
* easy to use  
  
All you need to do is,
1. download it from [here](https://github.com/ChinZJ/ip/releases/tag/A-Jar).  
1. double-click it.
1. add your tasks.  
1. let it manage your tasks for you. 🤙   
  
And it is *FREE*!  
  
Features:  
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders
- [ ] User friendly Graphical user interface  
  
---  
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
package botling;

import botling.commands.CommandParser;
import java.util.Scanner;

/**
 * Main class where application starts.
 */
public class Botling {
    /**
     * Generates objects required for program to function.
     * Handles the main loop of the program
     */
    public static void main(String[] args) {
        Scanner reader = new Scanner(System.in);
        TaskList tasks = new TaskList();
        CommandParser cmdParse = new CommandParser();

        // Generates the startup.
        System.out.println(cmdParse.start(tasks));

        // User input loop.
        while (tasks.isOpen()) {
            // Leveraging on object pass by reference
            System.out.println(cmdParse.parse(reader.nextLine(), tasks));
        }
        reader.close();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ChinZJ committed:
Renamed Chatbots and Greetings

The codebase contains Duke as a template placeholder.

The name of the chatbot needs to be personalized for our chatbot.

Let's,
* Remove all Duke references from the source code.
* Rename chatbot to Botling.
* Implement basic greeting and exit functionality.

Level-0 functionality implemented according to requirements.

ChinZJ committed:
Add Echo Functionality

Botling currently greets and exits.

Echo functionality serves as a foundation for basic user interaction.

The name of the chatbot needs to be personalized for our chatbot.

Let's,
* Echo commands from user input.
* Exit when "bye" is given as an input.

Level-1 functionality implemented according to requirements.

ChinZJ committed:
Add List Functionality

Botling currently echoes commands.

List functionality serves as a foundation for tracking tasks.

Let's,
* Store all inputs by the user except for "list" and "bye".
* Display all inputs by the user when "list" is given as an input.

Level-2 functionality implemented according to requirements.

ChinZJ committed:
Add Task Class

Botling currently lists tasks.

Creating a class for tasks follows in accordance with OOP.
Allows for abstraction and better structure.

Let's,
* Create a Task class with two attributes - name and isDone.
* Each Task object is responsible for their own statuses (isDone).

Level-3 functionality implemented according to requirements.

ChinZJ committed:
Add Mark as Done Functionality

Botling currently has a new Task class.

This allows for tasks to be marked as done depending on their status.

Let's,
* Update Botling's main method to incorporate the mark as done features.

Level-3 functionality implemented according to requirements.

ChinZJ committed:
Reconfigured Task Constructor

Task class currently has two constructors.

Reconfiguring the first constructor to utilize the second constructor.
The first constructor is a more specific version of the second.

Let's,
* Update Task class first constructor to use the second constructor.

Level-3 functionality implemented according to requirements.

ChinZJ committed:
Add Task class : toString() method

ChinZJ committed:
Add ToDo class

Currently only Task class exists.

Different subtypes of Tasks have different outputs.

Let's,
* Create a ToDo class that extends Task class.

Level-4 functionality implemented according to requirements.

ChinZJ committed:
Add Deadlines class

Currently only Task and ToDo class exists.

Different subtypes of Tasks have different outputs.

Let's,
* Create a Deadlines class that extends Task class.

Level-4 functionality implemented according to requirements.

ChinZJ committed:
Add Events class

Currently only Task, ToDo and Deadlines class exists.

Different subtypes of Tasks have different outputs.

Let's,
* Create a Events class that extends Task class.

Level-4 functionality implemented according to requirements.

ChinZJ committed:
Add ToDos, Deadlines and Events functionality

Currently Task, ToDo Deadlines and Events class exists.

They have not been implemented into Botling yet.

Let's,
* Implement the functionality of the various classes in Botling main().

Level-4 functionality implemented according to requirements.

ChinZJ committed:
Add File to .gitignore

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add File to .gitignore

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add input.txt, EXPECTED.TXT and runtest.bat

Currently test cases are manually run.

Manual testing is prone to human errors and is time consuming.

Let's,
* Create a rigorous set of test cases for all current functionality.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add MsgConstants and ValConstants Class

Currently all constants are defined under Botling.java.

There is a long list of constants that should be defined under Enum.

Let's,
* Create a MsgConstants Enum class for String constants.
* Create a ValConstants Enum class for int constants.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add TaskList class

Currently all tasks are stored in an ArrayList under Botling.java

This does not provide task specific functionalities eg. marking a task.

Let's,
* Create a TaskList class which wraps the ArrayList<Task>.
* Add further functionality such as mark(), unmark().
* Print messages where applicable.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add MSG_EMPTY to MsgConstants.java

ChinZJ committed:
Redirect Print from Task.java and TaskList.java to Botling.java

Currently print statements are created on demand when needed.

This does not adhere to OOP principles.
Botling should be handling expected messages.

Lets,
* Switch all method outputs to String where applicable.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add InvalidInputException.java

Currently exceptions are all handled under if else control blocks.

Unexpected inputs should be handled by exceptions.

Lets,
* Create an InvalidInputException class.
* Create respective methods that would print handled errors.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Add CommandParser.java

Currently all commands are handled by Botling.java.

A separate class should be created to isolate the parsing of commands.

Lets,
* Create a CommandParser.java class that parses all user inputs.
* Allow the CommandParser to handle all exceptions.
* Return all correct messages to Botling.java.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Update Botling.java

Currently Botling.java is outdated with the other classes.

Lets,
* Update the code for Botling.java to utilize all created functionality.

Level-5 functionality implemented according to requirements.

ChinZJ committed:
Update input.txt and EXPECTED.txt for Delete Task Functionality

ChinZJ committed:
Updated MsgConstants.java and ValConstants.java for Delete Functionality

ChinZJ committed:
Add Delete Task Functionality to TaskList.java

Currently TaskList can only add tasks.

Removing completed tasks should be natural for a tasks list.

Lets,
* Add a remove() method for TaskList class.

Level-6 functionality implemented according to requirements.

ChinZJ committed:
Add delete Functionality to CommandParser.java

Level-6 functionality implemented according to requirements.

ChinZJ committed:
Add JavaDocs for all methods

ChinZJ committed:
Refactor MsgConstants into MsgGenConst and CmdConst

Current MsgConstants contains all String constants.

However, different class uses different parts of the constants.
Constants should be unique to the specific class.

Lets,
* Refactor MsgConstants into MsgGenConst and CmdConst.
* MsgGenConst will be for messages produced by Botling.
* CmdConst will be for commands, as well as their expected syntax.

ChinZJ committed:
Add MsgGen class

Currently messages are printed immediately when required.

This goes against OOP practice.

Lets,
* Create a MsgGen class to generate all messages required.

ChinZJ committed:
Refactor All Remaining Code

Currently all code in the various classes are outdated.

Lets,
* Refactor all code to utilize the new MsgGen class.
* Ensure that all messages are generated by MsgGen.
* Only Botling.java should have print functionality.

ChinZJ committed:
Refactor JavaDocs and Code Syntax

ChinZJ committed:
Update EXPECTED.TXT and runtest.bat

ChinZJ committed:
Add TaskListWriter.java

Currently there is no functionality for restoring previous tasks.

Creating a file to read and write tasks allows for stored progress.

Lets,
* Create a history.txt file to store progress for the user locally.

Level-7 functionality implemented according to requirements.

ChinZJ committed:
Add Read and Write Capabilities for Tasks

Currently TaskListWriter has been implemented.
However it is not being used by other classes.

Lets,
* Create a write version of the tasks for TaskListWriter to store.
* Ensure the write version is readable and recoverable.

Level-7 functionality implemented according to requirements.

ChinZJ committed:
Update .gitignore for text-ui-test

ChinZJ committed:
Merge branch 'branch-Level-7'

ChinZJ committed:
Update text-ui-test

Level-8 functionality implemented according to requirements.

ChinZJ committed:
Add Date Functionality and DeadlineDate and EventDate classes

Currently Deadline and Event classes only take in String.
Creating a subclass that can read dates allows for future functionality
such as sorting Tasks by dates.

However, they should be able to recognize dates as well.

Lets,
* Create a child class of Deadlines and Events that can read dates.
* Create a hasDate() method for all classes to distinguish dates.

Level-8 functionality implemented according to requirements.

ChinZJ committed:
Add DateParser class

Currently, no class can read dates.

Lets,
* Create a DateParser class to read dates.

Level-8 functionality implemented according to requirements.

ChinZJ committed:
Update All Relevant Classes

Currently, all classes have not incorporated the check for dates.

Lets,
* Update the classes respectively

Level-8 functionality implemented according to requirements.

ChinZJ committed:
Package All Files Respectively

Currently all files are under the source root file.

Lets,
* Move all files into their respective packages.

ChinZJ committed:
Merge pull request #1 from ChinZJ/A-Packages

A-Package

ChinZJ committed:
Merge Commit branch-Level-8

Updated all files from branch-Level-8 for merge commit.

ChinZJ committed:
Merge branch 'branch-Level-8'

ChinZJ committed:
Update runtest.bat to Accommodate for Packages

ChinZJ committed:
Merge branch 'add-gradle-support'

ChinZJ committed:
Edit build.gradle for Botling

ChinZJ committed:
Merge pull request #3 from ChinZJ/A-Gradle

A-Gradle

ChinZJ committed:
Adhere to Coding Standards

ChinZJ committed:
Add JUnit Tests

Currently tests are done via text-ui-test.

However that does not allow for class specific testing.

Let's,
* Create JUnit tests for core public methods of the program.

ChinZJ committed:
Merge pull request #4 from ChinZJ/A-JUnit

A-JUnit

ChinZJ committed:
Remove Capitalization in Package

ChinZJ committed:
Updated runtest.bat and EXPECTED.TXT Accordingly

ChinZJ committed:
Updated Configurations for A-Jar

ChinZJ committed:
Merge pull request #5 from ChinZJ/A-Jar

A-Jar

ChinZJ committed:
Removed unknownList() From MsgGen class.

Currently Two Methods Overlap: unknownList() and unknownCmd().

Lets,
* Remove unknownList().
* Replace unknownList() with unknownCmd().

ChinZJ committed:
Add find Command

Currently Botling has no find functionality.

It may be difficult to look for a subset of tasks.

Lets,
* Create a "find" command for users to search their tasks.
* Code it such that the find command is case insensitive.

ChinZJ committed:
Refactor list() in TaskList class

Currently list() method in TaskList class has an unnecessary if else block.

This creates unnecessary inefficiencies.

Lets,
* Remove the if else block.

ChinZJ committed:
Refactored Code to Match Coding Standards

ChinZJ committed:
Merge branch 'branch-A-CodingStandard'

ChinZJ committed:
Merge branch 'branch-Level-9'

ChinZJ committed:
Update JUnit Test Cases to be Hardcoded

ChinZJ committed:
Clean Up Code

ChinZJ committed:
Fix Bug in TaskListWriter class

Buffered reader in TaskListWriter class was not closed.

This caused issues when trying to delete the `history.txt` file
during other test cases.

Bug found by seniors on GitHub issues comment.

ChinZJ committed:
Add GUI to Botling

Currently Botling is interacted via the CLI.

It is not user friendly and does not have a visually appealing display.

Lets,
* Follow the Java FX guide to create a GUI for Botling

Level-10 requirements.

ChinZJ committed:
Change Task to abstract class

ChinZJ committed:
Add Checkstyle Configurations

ChinZJ committed:
Clean Code to Adhere to Checkstyle Configurations

ChinZJ committed:
Merge branch 'A-CheckStyle'

ChinZJ committed:
Update JUnit Test

ChinZJ committed:
Merge branch 'Level-10'

ChinZJ committed:
Change System Exit Upon Corrupt File to RuntimeException

Currently program exits if user chooses not to delete corrupt file.

This complicates testing issues.
It is also more intuitive to throw an error to catch the message.
In future, this will come in handy for UI reloading.

ChinZJ committed:
Include history.txt in gitignore

ChinZJ committed:
Change corrupt history.txt file handling

Currently history.txt file corrupt handling is handled by CLI.

However, that is a known error and should be handled in the GUI.

Lets,
* Update the functionality of TaskListWriter class to only handle
  recreating of the file.
* Move the message generation to CommandParser class.
* Change the user input to be handled in the GUI.

Note that currently if the user selects 'n', they can either press 'y'
to reset or 'bye' to reload the bot.

ChinZJ committed:
Convert Label to TextArea

Currently GUI for replies are Label.

The text is not copyable and is hard to copy and paste text when needed.

Let's,
* Change the Label to a TextArea

Note: Functionality incomplete.

ChinZJ committed:
Temporary Backup

ChinZJ committed:
Add new command and new class

Help command to list all commands.
New class CommandColor to try and see if TextArea can have multiple colors.

ChinZJ committed:
Hardcode TextFlow object over TextArea object

Works for the most part, but special characters such as '}' are not
aligned.

Also note that String.format() can easily be broken with '%X' where X
is any non parseable character.

ChinZJ committed:
Add Selectable with Text Color Functionality to GUI

Underlying implementation comprises of a stacked TextArea object with
TextFlow object.
Note that the TextArea object is slightly misaligned, but was done
to the best of my capabilities at this moment.
TextArea object is made invisible, while TextFlow object will be colored.

ChinZJ committed:
Merge branch 'A-Extra-TextArea' into A-Extra

ChinZJ committed:
Merge branch 'A-Extra'

ChinZJ committed:
Refactor Selectable Text

Created custom class SelectableTextFlow

ChinZJ committed:
Merge branch 'A-Extra-TextArea'

ChinZJ committed:
Update header comments

ChinZJ committed:
Add assertions

Currently, code does not have assertion checks.

Assertion checks are crucial in ensuring programmer code runs smoothly.

Let's,
* Include assertion checks where appropriate.

ChinZJ committed:
Refactor commands

Currently all commands are being parsed by CommandParser class.

This results in very long code within methodology and is unpleasant.

Let's,
* Refactor the code and offset all work to their respective commands.

ChinZJ committed:
Refactor Command class to have child class TasksCmd

Currently Command class has all commands.

However some commands, particularly the tasks commands have
specialised tasks such as parsing inputs from the data file.

Let's,
* Extend the Command class with a TasksCmd class.
* Refactro all task related commands.

ChinZJ committed:
Merge pull request #7 from ChinZJ/branch-A-Assertions

branch-A-Assertions

ChinZJ committed:
Merge branch 'master' into branch-A-CodeQuality

ChinZJ committed:
Merge pull request #6 from ChinZJ/branch-A-CodeQuality

branch-A-CodeQuality

ChinZJ committed:
Fix send button bug

Directly accessing the vertical scrolls bar caused the button to throw
errors, presumably when loading the pane.

ChinZJ committed:
Add GitHub Action file

ChinZJ committed:
Merge branch 'A-CI'

ChinZJ committed:
Remove IO Redirection in GitHub Actions

ChinZJ committed:
Add sort capability to task list

Currently task list sorts naturally by chronological order.

However, tasks with dates are more likely to be important.

Let's,
* Create a custom sort functionality prioritizing end date, start and
  lastly chronological date.

ChinZJ committed:
Merge branch 'C-sort'

gengyudong commented on src/main/java/botling/DateParser.java:
```diff
@@ -0,0 +1,63 @@
+public class DateParser {
+    private static final Map<String, DateTimeFormatter> DATE_MAPPER;
+
+    static {
+        Map<String, DateTimeFormatter> tempMap = new HashMap<>();
+        tempMap.put("yy-MM-dd", DateTimeFormatter.ofPattern("yy-MM-dd"));
+        tempMap.put("yy-MM-dd HHmm", DateTimeFormatter.ofPattern("yy-MM-dd HHmm"));
+        tempMap.put("yyyy-MM-dd", DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+        tempMap.put("yyyy-MM-dd HHmm", DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+        tempMap.put("dd/MM/yyyy", DateTimeFormatter.ofPattern("dd/MM/yyyy"));
+        tempMap.put("dd/MM/yyyy HHmm", DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm"));
+        tempMap.put("dd/MM/yy", DateTimeFormatter.ofPattern("dd/MM/yy"));
+        tempMap.put("dd/MM/yy HHmm", DateTimeFormatter.ofPattern("dd/MM/yy HHmm"));
+        tempMap.put("dd MMM yyyy", DateTimeFormatter.ofPattern("dd MMM yyyy"));
+        tempMap.put("dd MMM yyyy HHmm", DateTimeFormatter.ofPattern("dd MMM yyyy HHmm"));
+        tempMap.put("dd MMM yy", DateTimeFormatter.ofPattern("dd MMM yy"));
+        tempMap.put("dd MMM yy HHmm", DateTimeFormatter.ofPattern("dd MMM yy HHmm"));
+        DATE_MAPPER = Collections.unmodifiableMap(tempMap);
+    }
```
I really like this part where you took into account different DateTime formats!

ChinZJ replied:
Thank you!

gengyudong commented on src/main/java/botling/TaskList.java:
```diff
@@ -0,0 +1,159 @@
+    private static final boolean TASK_IS_DONE = true;
+
+    private ArrayList<Task> tasks;
+    private boolean isActive;
```
Would it be more efficient to initialize the 'isActive' field to be true as compared to declaring the 'TASK_IS_DONE' variable and then assigning it to the 'isActive' field in the constructor?

ChinZJ replied:
Not too sure if setting true without that would be considered a magic number of sorts? What do you think.

gengyudong commented on src/main/java/botling/DateParser.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Check if user input is valid as a date object.
+ */
+public class DateParser {
```
Good naming convention - setting class name to be in verb form!

ChinZJ replied:
Thank you!

gengyudong commented on src/main/java/botling/commands/commandtypes/DeadlineCmd.java:
```diff
@@ -0,0 +1,79 @@
+    private static final int IDX_NAME = 0;
+    private static final int IDX_BY = 1;
+
+    /**
+     * Creates a deadline task.
+     */
```
Would it be clearer to include the params of this method in the Javadoc comments?

ChinZJ replied:
Ah yes, forgot about it completely, thank you for the reminder.

gengyudong reviewed (commented):
Overall, good job in following the coding standards! Your coding skills are really good! Keep up the great work!

ChinZJ committed:
Disable userinput upon shutdown

ChinZJ committed:
Add description to method Javadocs

ChinZJ committed:
Add name to application startup

ChinZJ committed:
Merge branch 'A-BetterGUI'

ChinZJ committed:
Add flair to Botling outputs

ChinZJ committed:
Merge branch 'A-Personality'

ChinZJ committed:
Add feature to detect dates of correct format but are invalid

ChinZJ committed:
Merge branch 'A-MoreErrorHandling'

ChinZJ committed:
Add more JUnit test to all classes except Command classes

All command classes are tested by the CommandParser class.

ChinZJ committed:
Add more JUnit testing

Exception of Command classes as they have been extensively tested
in CommandParser class.

ChinZJ committed:
Merge branch 'A-MoreTesting'

ChinZJ committed:
Add image citations and demo image

ChinZJ committed:
Update README.md

ChinZJ committed:
Merge branch 'A-UserGuide'

ChinZJ committed:
Update README.md

ChinZJ committed:
Create _config.yml

ChinZJ committed:
Update _config.yml

ChinZJ committed:
Update README.md

ChinZJ committed:
Change UI settings for selectable-text-flow.css

ChinZJ committed:
Update README.md

ChinZJ committed:
Update README.md

ChinZJ committed:
Update README.md

ChinZJ committed:
Update README.md

ChinZJ committed:
Update README.md

ChinZJ committed:
Update README.md

ChinZJ committed:
Change parsing of help, list, find mark, unmark and delete task

Bye, help and list commands were previously space sensitive.
Mark, unmark and delete commands previously throws invalid format for
any unexpected cases.

Now bye, help and list are insensitive to spaces added after the
command.
Now mark, unmark and delete will prioritize throwing a no
task error message instead.

ChinZJ committed:
Refactor JUnit test to adhere to SLAP

ChinZJ committed:
Add stripping function to leading white spaces for user inputs
